### PR TITLE
"http" backend to return existing lock metadata on lock conflict, rather than new lock metadata

### PR DESCRIPTION
**Pull request by haiderbari**
Resolves # Bug 2004 https://github.com/opentofu/opentofu/issues/2004

## Target Release

1.9.0, if possible back port

## Checklist

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.
- [x] I (and other contributors to this PR) have not looked at the Terraform source code while implementing this PR.

### Go checklist

 

- [x] I have run golangci-lint on my change and receive no errors relevant to my code.
- [x] I have run existing tests to ensure my code doesn't break anything.
- [x] I have added tests for all relevant use cases of my code, and those tests are passing.
- [x] I have only exported functions, variables and structs that should be used from other packages.
- [x] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

- [ ] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

### CODE REVIEW

File: internal/backend/remote-state/http/client_test.go [hunk 1]
```diff
@@ -251,3 +255,136 @@ func TestHttpClient_IsLockingEnabled(t *testing.T) {
...
+		return space.ReplaceAllString(str, " ")
+	}
+
+	testCases := []struct {
+		name           string
+		lockMethod     string
+		lockInfo       *statemgr.LockInfo
+		handler        http.HandlerFunc
+		validateResult func(lockID string, errorMessage error)
+	}{
```
Comment (ollevche) [inline internal/backend/remote-state/http/client_test.go:281]: I would prefer a more clear separation between test logic and test data (inputs and outputs). I think the following structure covers test needs in a slightly simpler way (you would also need to change test run logic inside a loop):

```suggestion
	inputLockInfo := statemgr.LockInfo{
		ID:        "ada-lovelace-state-lock-id",
		Who:       "AdaLovelace",
		Operation: "TestTypePlan",
	}

	testCases := []struct {
		name				string
		lockResponseStatus	int
		lockResponseBody		[]byte
		expectedLockInfo		*statemgr.LockInfo // or just an ID
		expectedErrorMsg		string
	}{
	...
```

File: internal/backend/remote-state/http/client_test.go [hunk 2]
```diff
@@ -251,3 +255,136 @@ func TestHttpClient_IsLockingEnabled(t *testing.T) {
...
+					t.Fatalf("Lock, expected locked: ID= to be %q, got %q", expectedStateLockID, errorMessage.Error())
+				}
+
+				if !strings.Contains(errString, fmt.Sprintf("ID: %s", expectedStateLockID)) {
+					t.Fatalf("Lock, expected ID: to be %q, got %q", expectedStateLockID, errorMessage.Error())
+				}
+
+				if !strings.Contains(errString, "Who: LinusTorvalds") {
+					t.Fatalf("Lock, expected Who: to be LinusTorvalds, got %q", errorMessage.Error())
+				}
```
Comment (ollevche) [inline internal/backend/remote-state/http/client_test.go:335]: Can we compare two strings (error messages) directly?

PS with my previous suggestion, this comparison can be moved to a shared test run logic (inside a loop).

**Review by ollevche — COMMENTED** [review]
Thank you for the contribution! I left a few suggestions to simplify tests.

File: internal/backend/remote-state/http/client_test.go (hunk 1 above)
Comment (haiderbari) [reply internal/backend/remote-state/http/client_test.go]: I have updated the tests base your suggestion

File: internal/backend/remote-state/http/client_test.go (hunk 2 above)
Comment (haiderbari) [reply internal/backend/remote-state/http/client_test.go]: I have updated the comparison checks

File: internal/backend/remote-state/http/client_test.go [hunk 3]
```diff
@@ -251,3 +253,93 @@ func TestHttpClient_IsLockingEnabled(t *testing.T) {
...
+			lockResponseStatus:  http.StatusLocked,
+			lockResponseBody:    []byte(`{"ID":"linus-torvalds-http-remote-state-lock-id", "Path": "", "Operation": "TestTypePlan", "Who": "LinusTorvalds", "Created": "2024-08-15T09:00:26.00Z" }`),
+			expectedStateLockID: "",
+			expectedErrorMsg: &statemgr.LockError{
+				Info: &statemgr.LockInfo{
+					ID:        "linus-torvalds-http-remote-state-lock-id",
+					Who:       "LinusTorvalds",
+					Operation: "TestTypePlan",
+					Created:   time.Date(2024, time.August, 15, 9, 0, 26, 0, time.UTC),
+				},
```
Comment (ollevche) [inline internal/backend/remote-state/http/client_test.go:297]: nit (could be ignored): we can move this as a separate variable and reuse it for json generation and in the error message.

File: internal/backend/remote-state/http/client_test.go [hunk 4]
```diff
@@ -251,3 +253,93 @@ func TestHttpClient_IsLockingEnabled(t *testing.T) {
...
+			client := &httpClient{
+				LockURL:    lockURL,
+				LockMethod: "LOCK",
+				Client:     retryablehttp.NewClient(),
+			}
+
+			lockID, err := client.Lock(tt.lockInfo)
+			if tt.expectedErrorMsg != nil && err.Error() != tt.expectedErrorMsg.Error() {
+				t.Errorf("Lock() error = %v, want %v", err, tt.expectedErrorMsg)
+			}
```
Comment (ollevche) [inline internal/backend/remote-state/http/client_test.go:339]: We must check for (1) unexpected error, (2) absence of expected error and (3) mismatched errors. Currently, we only check for (3) and panic on (2).

**Review by ollevche — CHANGES_REQUESTED** [review]

File: internal/backend/remote-state/http/client_test.go (hunk 3 above)
Comment (haiderbari) [reply internal/backend/remote-state/http/client_test.go:297]: thanks for the feedback, I have already implemented it, I haven't pushed that change yet.

File: internal/backend/remote-state/http/client_test.go (hunk 4 above)
Comment (ollevche) [reply internal/backend/remote-state/http/client_test.go:339]: I think we can do it without a separate variable:
```go
			lockID, err := client.Lock(tt.lockInfo)
			if tt.expectedErrorMsg != nil && err == nil {
				// no expected error
			}
			if tt.expectedErrorMsg == nil && err != nil {
				// unexpected error
			}
			if tt.expectedErrorMsg != nil && err != nil && err.Error() != tt.expectedErrorMsg.Error() {
				// mismatched errors
			}			
```

Comment (haiderbari) [reply internal/backend/remote-state/http/client_test.go:339]: I think this capture all use cases 

`if (err != nil) != (tt.expectedErrorMsg != nil) || (err != nil && err.Error() != tt.expectedErrorMsg.Error()) {
	// One of errors is nil and the other is not OR if both are not nil but their messages differ
}
`

Comment (ollevche) [reply internal/backend/remote-state/http/client_test.go:339]: I would prefer a simpler-to-read option with multiple ifs, it also allows us to create different error messages based on different cases.

Comment (haiderbari) [reply internal/backend/remote-state/http/client_test.go]: Hi @ollevche I have updated the tests base on your suggestions, thanks

File: internal/backend/remote-state/http/client_test.go (hunk 3 above)
Comment (haiderbari) [reply internal/backend/remote-state/http/client_test.go]: Moved it to it's own variable.

**Comment by ollevche** [comment]
Hey @haiderbari could you please also fix the linting issue?

**Comment by haiderbari** [comment]
> Hey @haiderbari could you please also fix the linting issue?

Hi @ollevche does the version matter for golanglint ci. When I ran the command it didn't flag that line for me or are you not using the default linting rules.

**Pull request merged**